### PR TITLE
implement token encoder

### PR DESCRIPTION
**Pull request by beesaferoot**

### CODE REVIEW

File: token/shared/shared.go [hunk 1]
```diff
@@ -96,16 +95,20 @@ func convertTo29BitsAndHalf(source uint64) uint32 {
...
-	for i := range 15 {
-		thisArray := bitArray[i*2 : (i*2)+2]
-		restrictedDigitToken.WriteString(
-			fmt.Sprint(bitArrayToInt(thisArray) + 1))
-	}
+	//for i := range 4 {
+	//	thisArray := bitArray[i*2 : (i*2)+2]
+	//	restrictedDigitToken.WriteString(
+	//		fmt.Sprint(bitArrayToInt(thisArray) + 1))
+	//}
```
Comment (dmohns) [inline token/shared/shared.go]: Is this commented out only temporary? Or could we remove it?

File: token/token_test.go [hunk 2]
```diff
@@ -1 +1,314 @@
...
+	StartingCode       uint32 `json:"starting_code"`
+	Key                string `json:"key"`
+	TokenCount         uint8  `json:"token_count"`
+	RestrictedDigitSet bool   `json:"restricted_digit_set"`
+	TokenType          int    `json:"token_type"`
+	ValueRaw           int    `json:"value_raw"`
+	Token              string `json:"token"`
+}
+
+func getTestTokens() ([]TTokenContext, error) {
```
Comment (dmohns) [inline token/token_test.go:29]: Would it be possible to read the file from disk? In the future I would like to update the test cases with even more and make sure all OpenPAYGO libraries use exactly the same file.

**Review by dmohns — COMMENTED** [review]
Looks great, few comments

File: token/shared/shared.go (hunk 1 above)
Comment (beesaferoot) [reply token/shared/shared.go]: I commented this part out as it didn't give the expected result from the tests. It turns out ```ir := bitArrayToInt(bitArray)``` was enough to get the function to work correctly. 

I will remove it if no further comments on its usefulness is added.

File: token/token_test.go (hunk 2 above)
Comment (beesaferoot) [reply token/token_test.go:29]: Yes, should be able to add this in I believe.

Comment (beesaferoot) [reply token/token_test.go:29]: Updated.

File: token/shared/shared.go (hunk 1 above)
Comment (beesaferoot) [reply token/shared/shared.go]: Updated.

**Comment by dmohns** [comment]
Perfect, thanks @beesaferoot 

One last request: I refactored the Github Actions setup a bit. Could you rebase (or merge) the latest changed from the `main` branch and resolve an merge conflicts?

**Comment by beesaferoot** [comment]
@dmohns Now updated.

Failing tests are due to the extended token gen not being implemented yet.

**Comment by beesaferoot** [comment]
Kindly hold off merging, I just realized some of the tests are failing due to commented out code from earlier (`ConvertTo4DigitsToken`)

Now fixed. 
@dmohns

**Comment by dmohns** [comment]
> @dmohns Now updated.
> 
> Failing tests are due to the extended token gen not being implemented yet.

Got it 👍 One last request here: Could you update (probably expand it) the table in the main README to make it more visible which features are implemented, and which is missing? I.e. "normal" tokens work, but "exdended" not yet?

**Comment by beesaferoot** [comment]
> Got it 👍 One last request here: Could you update (probably expand it) the table in the main README to make it more visible which features are implemented, and which is missing? I.e. "normal" tokens work, but "exdended" not yet?

Updated the README, kindly review.

**Review by dmohns — APPROVED** [review]
Thanks a lot!!

**Pull request merged**